server/blob/blob-joining/src/test/java/org/apache/james/blob/joining/JoiningBlobStoreTest.java

```diff
@@ -75,139 +60,211 @@ public InputStream read(BlobId blobId) {
         }
     }
 
-    private static final HashBlobId.Factory BLOB_ID_FACTORY = new HashBlobId.Factory();
-    private static final MemoryBlobStore PRIMARY_BLOB_STORE = new MemoryBlobStore(BLOB_ID_FACTORY);
-    private static final MemoryBlobStore SECONDARY_BLOB_STORE = new MemoryBlobStore(BLOB_ID_FACTORY);
-    private static final ThrowingBlobStore THROWING_BLOB_STORE = new ThrowingBlobStore();
+    private static class ThrowingBlobStore implements BlobStore {
+
+        @Override
+        public CompletableFuture<BlobId> save(byte[] data) {
+            throw new RuntimeException("not supported");
+        }
+
+        @Override
+        public CompletableFuture<BlobId> save(InputStream data) {
+            throw new RuntimeException("not supported");
+        }
+
+        @Override
+        public CompletableFuture<byte[]> readBytes(BlobId blobId) {
+            throw new RuntimeException("not supported");
+        }
+
+        @Override
+        public InputStream read(BlobId blobId) {
+            throw new RuntimeException("not supported");
+        }
+    }
 
+    private static final HashBlobId.Factory BLOB_ID_FACTORY = new HashBlobId.Factory();
     private static final byte [] BLOB_CONTENT = "blob content".getBytes();
-    private static final byte [] BLOB_CONTENT_CAUSES_THROWING_DIRECTLY = "blob content causes throwing directly".getBytes();
 
-    @AfterEach
-    void tearDown() {
-        PRIMARY_BLOB_STORE.clear();
-        SECONDARY_BLOB_STORE.clear();
+    private JoiningBlobStore joiningBlobStore;
+
+    @BeforeEach
+    void setup() {
+        MemoryBlobStore primary = new MemoryBlobStore(BLOB_ID_FACTORY);
+        MemoryBlobStore secondary = new MemoryBlobStore(BLOB_ID_FACTORY);
+        joiningBlobStore = new JoiningBlobStore(primary, secondary);
     }
 
     @Override
     public BlobStore testee() {
-        return new JoiningBlobStore(PRIMARY_BLOB_STORE, SECONDARY_BLOB_STORE);
+        return joiningBlobStore;
     }
 
     @Override
     public BlobId.Factory blobIdFactory() {
         return BLOB_ID_FACTORY;
     }
 
+    @Nested
+    class PrimarySaveThrowsExceptionDirectly {
```

Comment:
Can you also read from the `joining blob store` and ensure you can read from the secondary when the primary fails?